core/src/main/scala/org/apache/spark/api/r/RRunner.scala

```diff
@@ -44,380 +35,149 @@ private[spark] class RRunner[U](
+        }
       }
-    }
-  }
-
-  /**
-   * Start a thread to write RDD data to the R process.
-   */
-  private def startStdinThread(
-      output: OutputStream,
-      iter: Iterator[_],
-      partitionIndex: Int): Unit = {
-    val env = SparkEnv.get
-    val taskContext = TaskContext.get()
-    val bufferSize = System.getProperty(BUFFER_SIZE.key,
-      BUFFER_SIZE.defaultValueString).toInt
-    val stream = new BufferedOutputStream(output, bufferSize)
 
-    new Thread("writer for R") {
-      override def run(): Unit = {
+      /**
+       * Reads next object from the stream.
+       * When the stream reaches end of data, needs to process the following sections,
+       * and then returns null.
+       */
+      override protected def read(): OUT = {
         try {
-          SparkEnv.set(env)
-          TaskContext.setTaskContext(taskContext)
-          val dataOut = new DataOutputStream(stream)
-          dataOut.writeInt(partitionIndex)
-
-          SerDe.writeString(dataOut, deserializer)
-          SerDe.writeString(dataOut, serializer)
-
-          dataOut.writeInt(packageNames.length)
-          dataOut.write(packageNames)
-
-          dataOut.writeInt(func.length)
-          dataOut.write(func)
-
-          dataOut.writeInt(broadcastVars.length)
-          broadcastVars.foreach { broadcast =>
-            // TODO(shivaram): Read a Long in R to avoid this cast
-            dataOut.writeInt(broadcast.id.toInt)
-            // TODO: Pass a byte array from R to avoid this cast ?
-            val broadcastByteArr = broadcast.value.asInstanceOf[Array[Byte]]
-            dataOut.writeInt(broadcastByteArr.length)
-            dataOut.write(broadcastByteArr)
-          }
-
-          dataOut.writeInt(numPartitions)
-          dataOut.writeInt(mode)
-
-          if (isDataFrame) {
-            SerDe.writeObject(dataOut, colNames, jvmObjectTracker = null)
+          val length = dataStream.readInt()
+
+          length match {
+            case SpecialLengths.TIMING_DATA =>
+              // Timing data from R worker
+              val boot = dataStream.readDouble - bootTime
+              val init = dataStream.readDouble
+              val broadcast = dataStream.readDouble
+              val input = dataStream.readDouble
+              val compute = dataStream.readDouble
+              val output = dataStream.readDouble
+              logInfo(
+                ("Times: boot = %.3f s, init = %.3f s, broadcast = %.3f s, " +
+                  "read-input = %.3f s, compute = %.3f s, write-output = %.3f s, " +
+                  "total = %.3f s").format(
+                  boot,
+                  init,
+                  broadcast,
+                  input,
+                  compute,
+                  output,
+                  boot + init + broadcast + input + compute + output))
+              read()
+            case length if length > 0 =>
+              readData(length).asInstanceOf[OUT]
+            case length if length == 0 =>
```

Comment:
I tried to not add or remove existing logics but this logic was added to reuse `ReaderIterator`. The end of stream should be marked explicitly (this is consistent with Python side).

Reply:
I see this new case - how can we test this more?

Reply:
This is actually being tested by existing tests in R. (I am going to explain from scratch to completely make sure we're synced on this).

Here is when the data format is sent from R process to JVM via socket.


```
----------------------------
| ... # Output data        |
----------------------------
| -1 # Timing data start   |
| ... # Computed time, etc.|
----------------------------
| 0 # End of stream        |
----------------------------
```

**Before:**

Previously, RRunner's iterator in this code path stopped iteration when `read()` returns `null`. So, `readData` returned `null` when it's `0`.

**After:**

Now, RRunner needs to mark the end of stream explicitly via setting `eos` to `true` (This prevents when the returned value of `read()` is `null` and it has some meanings about it).

So, in other words, it works if we mark `eos` to `true` at everywhere it returned `null` in the previous RRunner, 

https://github.com/apache/spark/blob/8126d09fb5b969c1e293f1f8c41bec35357f74b5/core/src/main/scala/org/apache/spark/api/r/RRunner.scala#L290

https://github.com/apache/spark/blob/8126d09fb5b969c1e293f1f8c41bec35357f74b5/core/src/main/scala/org/apache/spark/api/r/RRunner.scala#L282

https://github.com/apache/spark/blob/8126d09fb5b969c1e293f1f8c41bec35357f74b5/core/src/main/scala/org/apache/spark/api/r/RRunner.scala#L272

However, explicitly marking the end of stream when 0 is read at one place should be better. So I decided to do it as so.

Reply:
In case or ArrowRRunner, it behaves same as previous.